server/src/main/java/org/elasticsearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -249,6 +249,115 @@ private static Object extractValue(String[] pathElements, int index, Object curr
+        } else if (suffixMaps.size() == 1) {
+            SuffixMap suffixMap = suffixMaps.getFirst();
+            suffixMap.map().put(suffixMap.suffix(), newValue);
+        } else {
+            throw new IllegalArgumentException(
+                "Path [" + path + "] could be inserted in " + suffixMaps.size() + " distinct ways, it is ambiguous which one to use"
+            );
+        }
+    }
+
+    private record SuffixMap(String suffix, Map<String, Object> map) {}
+
+    private static List<SuffixMap> extractSuffixMaps(String[] pathElements, int index, Object currentValue) {
+        if (currentValue instanceof List<?> valueList) {
+            List<SuffixMap> suffixMaps = new ArrayList<>(valueList.size());
+            for (Object o : valueList) {
+                suffixMaps.addAll(extractSuffixMaps(pathElements, index, o));
+            }
+
+            return suffixMaps;
+        } else if (currentValue instanceof Map<?, ?>) {
+            @SuppressWarnings("unchecked")
+            Map<String, Object> map = (Map<String, Object>) currentValue;
+            List<SuffixMap> suffixMaps = new ArrayList<>(map.size());
+
+            String key = pathElements[index];
+            while (index < pathElements.length) {
+                if (map.containsKey(key)) {
+                    if (index + 1 == pathElements.length) {
+                        // We found the complete path
+                        suffixMaps.add(new SuffixMap(key, map));
+                    } else {
+                        // We've matched that path partially, keep traversing to try to match it fully
+                        suffixMaps.addAll(extractSuffixMaps(pathElements, index + 1, map.get(key)));
+                    }
+                }
+
+                if (++index < pathElements.length) {
+                    key += "." + pathElements[index];
+                }
+            }
+
+            if (suffixMaps.isEmpty()) {
+                // We checked for all remaining elements in the path, and they do not exist. This means we found a leaf map that we should
+                // add the value to.
+                suffixMaps.add(new SuffixMap(key, map));
+            }
+
+            return suffixMaps;
+        } else {
+            // The path terminated early on a non-traversable data type. This is ok, it means we can't insert the value at this particular
+            // leaf node. Return an empty list to indicate this. The value will be inserted on the deepest appropriate parent map.
+            // This approach allows us to insert fields with dots in the name that share a common prefix, like so:
+            //
+            // Map<String, Object> map = new HashMap<>();
+            // XContentMapValues.insertValue("x.y", map, "value1");
+            // XContentMapValues.insertValue("x", map, "value2");
+            // XContentMapValues.insertValue("x.y.z", map, "value3");
+            return List.of();
```

Comment:
I don't think we should handle this. The example here would fail to index since `x` cannot be a value and an object at the same time. We should throw an exception like it was before this last change.

Reply:
This reminds me that a semantic_text field cannot be added on a path that sets `subobjects` to false. I guess that's what you try to handle here. However the semantic_text field is an object after the transformation so it won't work. We should instead disallow setting a semantic_text on a path that disable sub-objects.

Reply:
This could be a valid mapping if `subobjects: false` is set on the mapping root. Additionally, not handling this case leads to inconsistent behavior depending on the order in which the values are inserted:

```java
{
  // ok
  Map<String, Object> map = new HashMap<>();
  XContentMapValues.insertValue("x.y.z", map, "value3");
  XContentMapValues.insertValue("x.y", map, "value1");
  XContentMapValues.insertValue("x", map, "value2");
}

{
  // not ok
  Map<String, Object> map = new HashMap<>();
  XContentMapValues.insertValue("x.y", map, "value1");
  XContentMapValues.insertValue("x", map, "value2");
  XContentMapValues.insertValue("x.y.z", map, "value3");  <--- ERROR
}
```

This is a bad code smell to me and a fairly clear indicator we should handle this case. The downstream mapper validation can determine if the resulting map is actually a valid document, dependent on what subobjects are (or are not) allowed.

Reply:
@jimczi & I discussed this offline and we decided to remove handling of this case. However, not handling this case means that the `insertValue` method is less for general-purpose map modification and more specifically for `semantic_text` use cases. As such, we decided to move the method to `SemanticTextFieldMapper`.